--- a/net/minecraft/server/players/UserCache.java
+++ b/net/minecraft/server/players/UserCache.java
@@ -127,7 +127,7 @@
             usercache_usercacheentry.setLastAccess(this.getNextOperation());
             optional = Optional.of(usercache_usercacheentry.getProfile());
         } else {
-            optional = lookupGameProfile(this.profileRepository, s1);
+            optional = lookupGameProfile(this.profileRepository, s); // CraftBukkit - use correct case for offline players
             if (optional.isPresent()) {
                 this.add((GameProfile) optional.get());
                 flag = false;
@@ -198,7 +198,7 @@
             DateFormat dateformat = createDateFormat();
 
             jsonarray.forEach((jsonelement) -> {
-                Optional optional = readGameProfile(jsonelement, dateformat);
+                Optional<UserCache.UserCacheEntry> optional = readGameProfile(jsonelement, dateformat); // CraftBukkit - decompile error
 
                 Objects.requireNonNull(list);
                 optional.ifPresent(list::add);
