From 67a59fe522d5f08c97fe6c01edeaacf0efed9d26 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Fri, 28 Jun 2013 19:52:54 +1000
Subject: [PATCH] Disallow Interaction With Self


diff --git a/src/main/java/net/minecraft/server/network/PlayerConnection.java b/src/main/java/net/minecraft/server/network/PlayerConnection.java
index dff3d35f4..a4ad817e2 100644
--- a/src/main/java/net/minecraft/server/network/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/network/PlayerConnection.java
@@ -2514,6 +2514,13 @@ public class PlayerConnection extends ServerCommonPacketListenerImpl implements
         if (this.player.hasClientLoaded()) {
             final WorldServer worldserver = this.player.serverLevel();
             final Entity entity = packetplayinuseentity.getTarget(worldserver);
+            // Spigot Start
+            if ( entity == player && !player.isSpectator() )
+            {
+                disconnect( IChatBaseComponent.literal( "Cannot interact with self!" ) );
+                return;
+            }
+            // Spigot End
 
             this.player.resetLastActionTime();
             this.player.setShiftKeyDown(packetplayinuseentity.isUsingSecondaryAction());
-- 
2.49.0

