From 3e2c29ac8ba56614460db7855f340f58d66e899b Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Wed, 22 Jul 2015 19:04:37 +1000
Subject: [PATCH] Clear Packet Queue on Disconnect


diff --git a/src/main/java/net/minecraft/network/NetworkManager.java b/src/main/java/net/minecraft/network/NetworkManager.java
index 3b399bedf..855adebd4 100644
--- a/src/main/java/net/minecraft/network/NetworkManager.java
+++ b/src/main/java/net/minecraft/network/NetworkManager.java
@@ -672,6 +672,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
 
                     packetlistener1.onDisconnect(disconnectiondetails);
                 }
+                this.pendingActions.clear(); // Free up packet queue.
 
             }
         }
diff --git a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 96f280773..b4551ca7a 100644
--- a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -261,7 +261,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
     public void send(Packet<?> packet, @Nullable PacketSendListener packetsendlistener) {
         // CraftBukkit start
-        if (packet == null) {
+        if (packet == null || this.processedDisconnect) { // Spigot
             return;
         } else if (packet instanceof PacketPlayOutSpawnPosition) {
             PacketPlayOutSpawnPosition packet6 = (PacketPlayOutSpawnPosition) packet;
-- 
2.49.0

