--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -14,6 +14,16 @@
     public static final TicketType PORTAL = register("portal", 300L, true, TicketType.a.LOADING_AND_SIMULATION);
     public static final TicketType ENDER_PEARL = register("ender_pearl", 40L, false, TicketType.a.LOADING_AND_SIMULATION);
     public static final TicketType UNKNOWN = register("unknown", 1L, false, TicketType.a.LOADING);
+    // CraftBukkit start
+    public static long pluginTimeout = 0L;
+    public static final TicketType PLUGIN = register("plugin", 0L, false, TicketType.a.LOADING_AND_SIMULATION);
+    public static final TicketType PLUGIN_TICKET = register("plugin_ticket", 0L, false, TicketType.a.LOADING_AND_SIMULATION);
+
+    @Override
+    public long timeout() {
+        return (this != TicketType.PLUGIN) ? this.timeout : TicketType.pluginTimeout;
+    }
+    // CraftBukkit end
 
     private static TicketType register(String s, long i, boolean flag, TicketType.a tickettype_a) {
         return (TicketType) IRegistry.register(BuiltInRegistries.TICKET_TYPE, s, new TicketType(i, flag, tickettype_a));
