From 988426c05cfd38975c697637ef04903e627a49fd Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Sun, 2 Jun 2013 15:20:49 +1000
Subject: [PATCH] BungeeCord Support


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index da0ddce5..f26ffbdb 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2291,6 +2291,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     // Spigot start
     public class Spigot extends Entity.Spigot {
 
+        /**
+         * Gets the connection address of this player, regardless of whether it
+         * has been spoofed or not.
+         *
+         * @return the player's connection address
+         */
+        @NotNull
+        public InetSocketAddress getRawAddress() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
     }
 
     @NotNull
-- 
2.49.0

